File: src/main/java/com/soapdogg/lispInterpreter/tokenizer/Tokenizer.java
package com.soapdogg.lispInterpreter.tokenizer;

import java.util.List;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Scanner;

import com.soapdogg.lispInterpreter.datamodels.Token;
import com.soapdogg.lispInterpreter.asserter.LineFormatAsserter;
import lombok.AllArgsConstructor;

@AllArgsConstructor(staticName = "newInstance")
public class Tokenizer {

    private final ScannerToLineTransformer scannerToLineTransformer;
    private final LineFormatAsserter lineFormatAsserter;
    private final LineTokenizer lineTokenizer;

	public Queue<Token> tokenize(Scanner in) throws Exception {
		Queue<Token> tokens = new LinkedList<>();
        List<String> lines = scannerToLineTransformer.transformScannerInputToLines(
            in
        );
        for (String line : lines) {
            lineFormatAsserter.assertLineFormat(line);
            Queue<Token> tokenizedLineResult = lineTokenizer.tokenizeLine(
                line
            );
            tokens.addAll(tokenizedLineResult);
        }
        return tokens;
	}
}

File: src/main/java/com/soapdogg/lispInterpreter/tokenizer/Tokenizer.kt
package com.soapdogg.lispInterpreter.tokenizer

import com.soapdogg.lispInterpreter.asserter.LineFormatAsserter
import com.soapdogg.lispInterpreter.datamodels.Token
import java.util.Scanner

class Tokenizer(
    private val scannerToLineTransformer: ScannerToLineTransformer,
    private val lineFormatAsserter: LineFormatAsserter,
    private val lineTokenizer: LineTokenizer
) {

    fun tokenize(scanner: Scanner): List<Token> {
        val lines = scannerToLineTransformer.transformScannerInputToLines(
            scanner
        )
        lines.forEach { lineFormatAsserter.assertLineFormat(it) }
        return lines.map{
            line -> lineTokenizer.tokenizeLine(line)
        }.flatMap { it.toList() }
    }
}